### DISCUSSION TITLE
Use Timber in Theme and Plugin

### DISCUSSION THREAD

**Discussion by AfterglowWeb** [timber/timber#2815] [Q&A]
Hello there, I believe the question has been asked but I can't find any recent answer.
Our use case needs Twig templating in both theme and plugins.
The several Timber instances are conflicting on template directory and adding php functions to Twig.
My ideas are :
• filling a global with Timber object.
• making a minimal plugin with Timber and provide hooks so other components can use it.
Does anyone here would have a workaround idea please ?

**Comment by AfterglowWeb** [comment]
Start of an answer ; The only workaround I found is to leave Twig templating in the theme and only share datas between plugins and themes through custom WP filters and actions hooks. It is a consistant approach but not convenient at all if you are making plugins for unknown context. Some usefull Timber helpers are missing too of course. 
So I am still searching how to isolate multiple Timber instances in the Wordpress scope before moving to another solution. Anyone ?

**Comment by calebsmithdev** [comment]
Have you had any luck getting to work? I'm getting ready to dive into the same process and hoping to find some inspiration. We have about 5 sites with different themes and then 5-6 plugins that could potentially use a Twig partial or a few.

I'm thinking of just keeping it all in a shared utilities plugin that is already installed on every site, then maybe doing a `require-dev` in composer for the theme so at least the typesensing is there at that level, then it's removed when compiled for prod. That's about the best I can come up with so far.

**Comment by jasalt** [comment]
Interested also how this would be done sustainably. Hacked together a block plugin and it seems to work on a site with a basic theme based on Timber starter so far https://github.com/jasalt/glidejs-cf-block.

**Comment by wonkeythemonkey** [comment]
I know that there is an ongoing discussion within the WP team to introduce some sort of dependency manager for plugins and themes, but it hasn't gotten very far. In the meantime, you could try using a composer tool that namespaces your dependencies, as described in this [Pressidium article about plugin conflicts](https://pressidium.com/blog/wordpress-plugin-conflicts-how-to-prevent-composer-dependency-hell/)

One approach I've used before is including Timber in both the theme and the plugin or plugins, then adding a check before loading the library to see if it already exists. When I control the code in all of the relevant themes and plugins, it's pretty straightforward to add something like:

```
if ( ! class_exists( 'Timber' ) ) {
  include_once( 'path/to/timber/library.php' );
}
```

That avoids conflicts, but it won't fix mismatched Timber versions.

If you wanted to reduce duplicate code by doing something like @calebsmithdev's utility plugin idea, the [Plugin Dependencies header introduced in WordPress 6.5](https://make.wordpress.org/core/2024/03/05/introducing-plugin-dependencies-in-wordpress-6-5/) could help with load order.

**Reply by calebsmithdev** [reply]
Thanks for the callout on the Plugin Dependency! I totally forgot that was released and have looked for an excuse to use it. I'll give it a shot and see what happens 👍

**Comment by jasalt** [comment]
More recent discussion with some additional links https://github.com/timber/timber/discussions/3094.

**Comment by mmorris8** [comment]
Related, I have opened a discussion on PHP Dev itself to draft out this RFC and get it implemented.

https://github.com/michael-lloyd-morris/php-modules-rfc/blob/main/php-modules.md

The problem Timber faces beyond WordPress' lack of a composer integration (Way, way overdue IMO) is PHP's inability to deal with multiple versions of packages which given the reality of WordPress plugin dev is going to happen eventually - some plugin is going to want to use Timber 2 and not be updated for years when Timber 3 comes along and if there are any BC issues in it or Twig (which I assume would also be updated by then) then any site trying to use both plugins would be out of luck.

Having the PHP engine able to deal with this aspect of the problem would make the task easier for WP Core.  Although composer is intended to be updated from the Command line, it is possible to update it from an app - Drupal does it.

The proposal linked above is still in infancy and I need help getting it to completion. Beyond that, I think the PHP Dev list needs a little brigading by the WordPress community as they like to pretend we don't exist.
